[14.0][IMP] product_import: reorganize for inheritance to import by batch

florentx opened the pull request:
Some catalogs contain 10'000 references and import does not complete due to other tasks running at the same time, and locking table.

Splitting product import into individual jobs should mitigate this issue.

**[ EDIT ]**
This PR reorganizes code in order to easily be able to split import by batch with method `split()` of `queue_job` add-on, or any similar machinery.

simahawk commented on product_import/wizard/product_import.py:
```diff
@@ -211,23 +212,29 @@ def create_product(self, parsed_product, chatter_msg, seller=None):
+        This method is called from a queue job.
+        """
+        messgs = []
+        product = self.create_update_product(parsed_product, messgs, seller_id)
```
Sounds weird to have a public method called by the private one. It should be the other way around.
And why passing empty `messgs`?

simahawk commented on product_import/tests/test_product_import.py:
```diff
@@ -86,8 +86,11 @@ def test_get_company_id(self):
-        products = self.wiz_model._create_products(
-            self.parsed_catalog, seller=self.supplier
+        self.wiz_model._import_products(self.parsed_catalog, seller=self.supplier)
+        products = (
```
existing = ...search()
new = ...search() - existing

simahawk commented on product_import/wizard/product_import.py:
```diff
@@ -211,23 +212,29 @@ def create_product(self, parsed_product, chatter_msg, seller=None):
+        for product_vals in catalogue["products"]:
+            # One job per product
+            self.with_delay()._create_update_product(product_vals, seller.id)
         self._bdimport.post_create_or_update(catalogue, seller, doc_filename=filename)
```
Is it correct to call the post update here?
Also, now you have 3 methods to the import... why do you need them all?

florentx replied:
Refactored.

twalter-c2c reviewed (approved):
LGTM

simahawk commented on product_import/__manifest__.py:
```diff
@@ -13,9 +13,12 @@
         # OCA/edi
         "base_business_document_import",
+        # OCA/queue
+        "queue_job",
```
I've thought again about this... We should remove this dependency IMO.
In the simplest case we won't need it, in the more complex case we'll probably use the framework. Then, we can leverage the framework to the split. 
Additionally, there's now a new feature in queue_job which is called `split` and allows to.... split items in batches.
See https://github.com/OCA/queue/pull/658

simahawk commented on product_import/wizard/product_import.py:
```diff
@@ -213,33 +248,33 @@ def create_product(self, parsed_product, chatter_msg, seller=None):
+            if products[-1] is None:
+                products = [product for product in products if product]
+            # One job for x products (chunk of 40)
+            wiz.with_delay()._create_update_products(products, seller.id)
```
here I would simply call `wiz.create_update_products` that will call `_create_update_products` which can be overridden or you can rely on https://github.com/OCA/queue/blob/16.0/queue_job/models/base.py#L145

florentx committed:
[IMP] product_import: Reorganize for inheritance to import by batch

florentx committed:
[REF] product_import: do not use the env.context to configure import

twalter-c2c reviewed (approved):
LGTM

florentx committed:
[FIX] product_import: use update() instead of write()

It is needed in order to have a safe recomputation of fields of other models.
There was an issue in a third-party add-on which has a field recomputed
when a new 'product.supplierinfo' is added to 'seller_ids' of the product.
